tt_metal/tt_metal.cpp

```diff
@@ -691,8 +737,26 @@ std::shared_ptr<Buffer> CreateBuffer(const ShardedBufferConfig &config) {
 
 void DeallocateBuffer(Buffer &buffer) { buffer.deallocate(); }
 
+void AssignGlobalBufferToProgram(std::shared_ptr<Buffer> buffer, std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>> program) {
+    if (buffer->device()->sw_command_queues_.size() == 0) {
+        std::visit([&buffer] (auto&& p) {
+            using program_type = std::decay_t<decltype(p)>;
+            if constexpr (std::is_same_v<program_type, std::reference_wrapper<Program>>) {
+                p.get().add_global_buffer(buffer);
+            }
+            else {
+                p->add_global_buffer(buffer);
+            }
+        }, program);
+    }
+    else {
+        EnqueueAddBufferToProgram(buffer-> device()->command_queue(), buffer, program, false);
+    }
```

Comment:
Is the top branch an optimization to bypass the SW CQ? I think it's better to just push and forget about it than to implement bypass paths for a little bit of perf gain

Reply:
Update: Aditya explained that "buffer->device()->sw_command_queues_.size() == 0" means legacy/slow dispatch mode, let's make a method for that. And then call `EnqueueAddBufferToProgramImpl` to avoid duplicating the functionality

Reply:
Cleaned up all `tt_metal.cpp` that needed to be both FD and SD compatible. For both cases, we directly call the `Enqueue` API provided by the CQ. For SD mode, we're able to do this by creating a single SW CQ in passthrough mode.